src/main/scala/com/gu/commercial/display/AdTargetParam.scala

```diff
@@ -0,0 +1,228 @@
+  val name = "edition"
+
+  def from(editionId: String): Option[EditionParam] = SingleValue.fromRaw(editionId.take(3)) map (EditionParam(_))
+}
+
+case class KeywordParam(value: MultipleValues) extends AdTargetParam {
+  override val name = KeywordParam.name
+}
+
+object KeywordParam {
+  val name = "k"
+
+  def from(item: Content): Option[KeywordParam] =
+    MultipleValues.fromTags(item) { tag =>
+      tag.`type` == Keyword || tag.paidContentType.contains("Topic")
+    } map (KeywordParam(_))
+
+  def fromItemId(id: String): Option[KeywordParam] = MultipleValues.fromItemId(id) map (KeywordParam(_))
+
+  def from(section: Section): Option[KeywordParam] = MultipleValues.fromItemId(section.id) map (KeywordParam(_))
+
+  def from(tag: Tag): Option[KeywordParam] =
+    if (tag.`type` == Keyword) MultipleValues.fromItemId(tag.id) map (KeywordParam(_))
+    else None
+
+  def fromPath(path: String): Option[KeywordParam] =
+    MultipleValues.fromRawValues(Set(path.stripPrefix("/"))) map (KeywordParam(_))
+}
+
+case class ObserverParam(value: SingleValue) extends AdTargetParam {
+  override val name = ObserverParam.name
+}
+
+object ObserverParam {
+  val name = "ob"
+
+  def from(item: Content): Option[ObserverParam] =
+    SingleValue.fromRaw(if (item.tags.exists(_.id == "publication/theobserver")) "t" else "") map (ObserverParam(_))
+}
+
+case class PathParam(value: SingleValue) extends AdTargetParam {
+  override val name = PathParam.name
+}
+
+object PathParam {
+  val name = "url"
+
+  def apply(value: String): PathParam           = PathParam(SingleValue(value))
+  def fromItemId(id: String): Option[PathParam] = SingleValue.fromRaw(s"/${id.stripPrefix("/")}") map (PathParam(_))
+  def from(item: Content): Option[PathParam]    = fromItemId(item.id)
+  def from(section: Section): Option[PathParam] = fromItemId(section.id)
+  def from(tag: Tag): Option[PathParam]         = fromItemId(tag.id)
+}
+
+case class PlatformParam(value: SingleValue) extends AdTargetParam {
+  override val name = PlatformParam.name
+}
+
+object PlatformParam {
+  val name = "p"
+
+  def apply(value: String): PlatformParam = PlatformParam(SingleValue(value))
+}
+
+case class SeriesParam(value: MultipleValues) extends AdTargetParam {
+  override val name = SeriesParam.name
+}
+
+object SeriesParam {
+  val name = "se"
+
+  def from(item: Content): Option[SeriesParam] =
+    MultipleValues.fromTags(item) { tag =>
+      tag.`type` == Series || tag.paidContentType.contains("Series")
+    } map (SeriesParam(_))
+
+  def from(tag: Tag): Option[SeriesParam] =
+    if (tag.`type` == Series) MultipleValues.fromItemId(tag.id) map (SeriesParam(_))
+    else None
+}
+
+case class ShortUrlParam(value: SingleValue) extends AdTargetParam {
+  override val name = ShortUrlParam.name
+}
+
+object ShortUrlParam {
+  val name = "sh"
+
+  def from(item: Content): Option[ShortUrlParam] =
+    SingleValue.fromRaw(item.fields.flatMap(_.shortUrl).getOrElse("")) map (ShortUrlParam(_))
```

Comment:
Isn't this always going to return `Some(ShortUrlParam(_))` regardless of what's in the fields?